app/controllers/uploads_controller.rb

```diff
@@ -139,8 +138,14 @@ def validation_messages
         generic_requirement_message('These columns can only contain numeric values: ', validations)
       when ActiveRecord::Validations::UniquenessValidator
         generic_requirement_message('These columns should contain unique values: ', validations)
+      when "#{klass.name}Validator".safe_constantize
+        "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
+      else
+        validations.class.to_s
       end
-    end
+    end.flatten
```

Comment:
Basically, I think you can undo all these `requirements`-related changes and change this line to 
```suggestion
    end.compact
```
And you're good to go.  No having to deal with duplicated messages, blank bullet points or arcan 'ActiveRecord::Validations::WhateverValidator' bullet points appearing.

Reply:
To clarify you just need `.compact` instead of `.select(&:present?)` because the default response from a `case` statement is `nil`.  `.compact` removes `nil`s from an array.

Reply:
changed